### PR TITLE
feat(refactor): migrate retry/cache/rate limit providers

### PR DESCRIPTION
**Pull request by james-a-morris**
Work to migrate these files will require additional integration tests + I will remove duplicated code from the relayer

### CODE REVIEW

File: src/providers/cachedProvider.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+export class CacheProvider extends RateLimitedProvider {
+  public readonly getBlockByNumberPrefix: string;
+  public readonly getLogsCachePrefix: string;
+  public readonly callCachePrefix: string;
+  public readonly baseTTL: number;
+
+  constructor(
+    providerCacheNamespace: string,
+    readonly redisClient?: CachingMechanismInterface,
```
Comment (james-a-morris) [inline src/providers/cachedProvider.ts:19]: Note: we didn't have a `RedisClient` instance in the SDK but we do have `CachingMechanismInterface` which should be able to take a `RedisClient`

File: .github/workflows/build.yml
Comment (james-a-morris) [inline .github/workflows/build.yml:1]: This should've been on v18

File: src/providers/cachedProvider.ts (hunk 1 above)
Comment (daywiss) [reply src/providers/cachedProvider.ts:14]: whats the idea behind this interface?

Comment (daywiss) [reply src/providers/cachedProvider.ts:14]: also assuming this is mainly copied from the other repo?

File: src/providers/rateLimitedProvider.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import { QueueObject, queue } from "async";
+import { ethers } from "ethers";
+import { RateLimitTask } from "./utils";
+import { getOriginFromURL } from "../utils/NetworkUtils";
+import { Logger } from "winston";
+import { Logger as umaLogger } from "@uma/logger";
+
+// This provider is a very small addition to the StaticJsonRpcProvider that ensures that no more than `maxConcurrency`
+// requests are ever in flight. It uses the async/queue library to manage this.
+export class RateLimitedProvider extends ethers.providers.StaticJsonRpcProvider {
```
Comment (daywiss) [inline src/providers/rateLimitedProvider.ts:14]: also copied?

File: src/providers/retryProvider.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
+import { ethers, logger } from "ethers";
+import { CachingMechanismInterface } from "../interfaces";
+import { delay, isDefined, isPromiseFulfilled, isPromiseRejected } from "../utils";
+import { getOriginFromURL } from "../utils/NetworkUtils";
+import { CacheProvider } from "./cachedProvider";
+import { compareRpcResults, createSendErrorWithMessage, formatProviderError } from "./utils";
+import { AugmentedLogger } from "@uma/logger";
+import { PROVIDER_CACHE_TTL } from "./constants";
+
+export class RetryProvider extends ethers.providers.StaticJsonRpcProvider {
```
Comment (daywiss) [inline src/providers/retryProvider.ts:10]: also copied?

**Review by daywiss — APPROVED** [review]
looks good just curious where any logical changes were if any

File: src/providers/cachedProvider.ts (hunk 1 above)
Comment (james-a-morris) [reply src/providers/cachedProvider.ts:14]: THe `CachingMechanismInterface` is a way the sdk can stay un-opinionated about the actual persistence layer while still serving basic get/set functionality. Think: Redis as a cache or Arweave as a cache (which we do in the relayer)

File: src/providers/rateLimitedProvider.ts (hunk 2 above)
Comment (james-a-morris) [reply src/providers/rateLimitedProvider.ts:14]: Yep this is copied

File: src/providers/retryProvider.ts (hunk 3 above)
Comment (james-a-morris) [reply src/providers/retryProvider.ts:10]: Correct

File: src/providers/cachedProvider.ts (hunk 1 above)
Comment (james-a-morris) [reply src/providers/cachedProvider.ts:14]: Yes, all 3 classes have been copied - the corresponding relayer-v2 PR removes the logic

**Review by daywiss — APPROVED** [review]

File: src/providers/utils.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import lodash from "lodash";
+import { compareArrayResultsWithIgnoredKeys, compareResultsAndFilterIgnoredKeys } from "../utils/ObjectUtils";
+
+export { lodash };
+
+/**
+ * A record of error codes that correspond to fields that should be ignored when comparing RPC results.
+ * This is used to compare results from different providers that may have different, but still valid, results.
+ */
+const IGNORED_ERROR_CODES = {
```
Comment (mrice32) [inline src/providers/utils.ts:13]: These aren't error codes, they are fields where we ignore their values in comparison, so I think we should update the name to something more like IGNORED_FIELDS or similar.

It's a good idea to pull these out, but I would've preferred to see this type of refactoring in a follow-up PR so it's easy to tell what changed.

File: src/providers/rateLimitedProvider.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    // This sets up the queue. Each task is executed by calling the superclass's send method, which fires off the
+    // request. This queue sends out requests concurrently, but stops once the concurrency limit is reached. The
+    // maxConcurrency is configured here.
+    this.queue = queue(async ({ sendArgs, resolve, reject }: RateLimitTask, callback: () => void) => {
+      await this.wrapSendWithLog(...sendArgs)
+        .then(resolve)
+        .catch(reject);
+      // we need this for the queue to know that the task is done
+      // @see: https://caolan.github.io/async/v3/global.html
+      callback();
```
Comment (mrice32) [inline src/providers/rateLimitedProvider.ts:37]: Are you sure nothing else changed with the upgrade to async? Do we have any simple tests to validate that this still works? Note: I would be okay with just a simple test to just validate our assumption that async works the way we expect. We don't have to test the full provider.

**Review by mrice32 — COMMENTED** [review]
Looks great! Two minor comments

File: src/providers/utils.ts (hunk 4 above)
Comment (james-a-morris) [reply src/providers/utils.ts]: Pulled out and placed in: https://github.com/across-protocol/sdk/pull/695

File: src/providers/utils.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
+// The async/queue library has a task-based interface for building a concurrent queue.
+
+import { providers } from "ethers";
+import lodash from "lodash";
+import { compareArrayResultsWithIgnoredKeys, compareResultsAndFilterIgnoredKeys } from "../utils/ObjectUtils";
+
+export { lodash };
```
Comment (pxrl) [inline src/providers/utils.ts:7]: I think we only use relatively few things from lodash here - is it worth only exporting the few things that we need?

File: src/utils/ObjectUtils.ts [hunk 7]
```diff
@@ -75,3 +79,42 @@ export function groupObjectCountsByProp(objects: any[], getProp: (obj: any) => s
...
+}
+
+export function compareArrayResultsWithIgnoredKeys(ignoredKeys: string[], objA: unknown[], objB: unknown[]): boolean {
+  // Remove ignored keys from each element of copied arrays.
+  const filteredA = objA?.map((obj) => deleteIgnoredKeys(ignoredKeys, obj as Record<string, unknown>));
+  const filteredB = objB?.map((obj) => deleteIgnoredKeys(ignoredKeys, obj as Record<string, unknown>));
+
+  // Compare objects without the ignored keys.
+  return isDefined(filteredA) && isDefined(filteredB) && lodash.isEqual(filteredA, filteredB);
+}
```
Comment (pxrl) [inline src/utils/ObjectUtils.ts:120]: Do these have general usefulness outside of the providers? If not then should we just drop them in `src/providers/utils` ?

File: src/providers/rateLimitedProvider.ts (hunk 5 above)
Comment (james-a-morris) [reply src/providers/rateLimitedProvider.ts:37]: Tests for both the queue with a callback and one without: https://github.com/across-protocol/sdk/pull/693/commits/35d2ceda2905f5d911013d799ddadab0b5163e3a

File: src/providers/rateLimitedProvider.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+  override send(method: string, params: Array<unknown>): Promise<unknown> {
+    // This simply creates a promise and adds the arguments and resolve and reject handlers to the task.
+    return new Promise<unknown>((resolve, reject) => {
+      const task: RateLimitTask = {
+        sendArgs: [method, params],
+        resolve,
+        reject,
+      };
+      void this.queue.push(task);
```
[the remark this answers is not in the document]
Comment (james-a-morris) [reply src/providers/rateLimitedProvider.ts:91]: https://github.com/across-protocol/sdk/pull/693/commits/57a6487ec941192292266774e226b7c12c8f3233

File: src/utils/ObjectUtils.ts (hunk 7 above)
Comment (james-a-morris) [reply src/utils/ObjectUtils.ts]: https://github.com/across-protocol/sdk/pull/693/commits/75528faa95972ff871ff0f36f28fafe7c814165e

File: src/providers/utils.ts (hunk 6 above)
Comment (james-a-morris) [reply src/providers/utils.ts]: https://github.com/across-protocol/sdk/pull/693/commits/75528faa95972ff871ff0f36f28fafe7c814165e

**Review by pxrl — APPROVED** [review]
LGTM; I'll defer to Matt for a final review.

**Review by mrice32 — APPROVED** [review]
LGTM!

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**